web-common/src/features/entity-management/resource-invalidations.ts

```diff
@@ -0,0 +1,136 @@
+import { ResourceKind } from "@rilldata/web-common/features/entity-management/resource-selectors";
+import { resourcesStore } from "@rilldata/web-common/features/entity-management/resources-store";
+import {
+  getRuntimeServiceGetResourceQueryKey,
+  getRuntimeServiceListResourcesQueryKey,
+  V1GetResourceResponse,
+  V1Resource,
+} from "@rilldata/web-common/runtime-client";
+import type { V1WatchResourcesResponse } from "@rilldata/web-common/runtime-client";
+import {
+  invalidateMetricsViewData,
+  invalidateProfilingQueries,
+  invalidationForMetricsViewData,
+} from "@rilldata/web-common/runtime-client/invalidation";
+import { isProfilingQuery } from "@rilldata/web-common/runtime-client/query-matcher";
+import { runtime } from "@rilldata/web-common/runtime-client/runtime-store";
+import type { QueryClient } from "@tanstack/svelte-query";
+import { get } from "svelte/types/runtime/store";
+
+const MainEntities: {
+  [kind in ResourceKind]?: boolean;
+} = {
+  [ResourceKind.ProjectParser]: false,
+  [ResourceKind.Source]: true,
+  [ResourceKind.Model]: true,
+  [ResourceKind.MetricsView]: true,
+};
+
+export function invalidateResourceResponse(
+  queryClient: QueryClient,
+  res: V1WatchResourcesResponse
+) {
+  // only process for the `ResourceKind` present in `MainEntities`
+  if (!(res.name.kind in MainEntities)) return;
+
+  const instanceId = get(runtime).instanceId;
+  // invalidations will wait until the re-fetched query is completed
+  // so, we should not `await` here
+  switch (res.event) {
+    case "RESOURCE_EVENT_WRITE":
+      invalidateResource(queryClient, instanceId, res.resource);
+      break;
+
+    case "RESOURCE_EVENT_DELETE":
+      invalidateRemovedResource(queryClient, instanceId, res.resource);
+      break;
+  }
+
+  // only re-fetch list queries for kinds in `MainEntities` and is ture
+  if (MainEntities[res.name.kind]) {
+    resourcesStore.setResource(res.resource);
+    queryClient.refetchQueries(
+      // we only use individual kind's queries
+      getRuntimeServiceListResourcesQueryKey(instanceId, {
+        kind: res.name.kind,
+      })
+    );
+  } else if (
+    res.name.kind === ResourceKind.ProjectParser &&
```

Comment:
On line 34, you're returning early if `ResourceKind.ProjectParser`

Reply:
I am checking if the kind is present in the map not its boolean value in 34. Lemme know if the comment doesnt make it clear.

Reply:
Ah whoops I misread that. If I understand this right, maybe this would be more clear:
```
const MainEntities = [ResourceKind.Source, ResourceKind.Model, ResourceKind.MetricsView]

...

if (res.name.kind in MainEntities || res.name.kind === ResourceKind.ProjectParser)
```